src/dpo/src/Optdp.cpp

```diff
@@ -0,0 +1,855 @@
+
+////////////////////////////////////////////////////////////////
+Optdp::~Optdp() {}
+
+////////////////////////////////////////////////////////////////
+void Optdp::init(ord::OpenRoad* openroad) {
+  openroad_ = openroad;
+  db_ = openroad->getDb();
+  logger_ = openroad->getLogger();
+}
+
+////////////////////////////////////////////////////////////////
+void Optdp::improvePlacement() {
+  logger_->report("Detailed placement improvement.");
+
+  dpl::Opendp* opendp = openroad_->getOpendp();
+  hpwlBefore_ = opendp->hpwl();
+
+  import();
+
+  // A manager to track cells.
+  dpo::DetailedMgr mgr(arch_, nw_, rt_);
+  mgr.setLogger(logger_);
+
+  // Legalization.  Doesn't particularly do much.  It only
+  // populates the data structures required for detailed
+  // improvement.  If it errors or prints a warning when
+  // given a legal placement, that likely means there is
+  // a bug in my code somewhere.
+  dpo::ShiftLegalizerParams lgParams;
+  dpo::ShiftLegalizer lg(lgParams);
+  lg.legalize(mgr);
+
+  // Detailed improvement.  Runs through a number of different
+  // optimizations aimed at wirelength improvement.  The last
+  // call to the random improver can be set to consider things
+  // like density, displacement, etc. in addition to wirelength.
+
+  dpo::DetailedParams dtParams;
+  dtParams.m_script = "";
+  // Maximum independent set matching.
+  dtParams.m_script += "mis -p 10 -t 0.005";
+  dtParams.m_script += ";";
+  // Global swaps - commented out right now.  caused padding violation???
+  //dtParams.m_script += "gs -p 10 -t 0.005";
+  //dtParams.m_script += ";";
+  // Vertical swaps - commented out right now.  caused padding violation???
+  //dtParams.m_script += "vs -p 10 -t 0.005";
+  //dtParams.m_script += ";";
+  // Small reordering.
+  dtParams.m_script += "ro -p 10 -t 0.005";
+  dtParams.m_script += ";";
+  // Random moves and swaps with hpwl as a cost function.
+  // Commented out moves generated with global and vertical swaps 
+  // since they seem to be causing padding violations.
+  dtParams.m_script +=
+       "default -p 5 -f 20 -gen rng -obj hpwl -cost (hpwl)";
+  dpo::Detailed dt(dtParams);
+  dt.improve(mgr);
+
+  // Write solution back.
+  updateDbInstLocations();
+
+  hpwlAfter_ = opendp->hpwl();
+
+  double dbu_micron = db_->getTech()->getDbUnitsPerMicron();
+
+  // Statistics.
+  logger_->report("Detailed Improvement Results");
+  logger_->report("------------------------------------------");
+  logger_->report("Original HPWL         {:10.1f} u", (hpwlBefore_/dbu_micron));
+  logger_->report("Final HPWL            {:10.1f} u", (hpwlAfter_/dbu_micron));
+  double hpwl_delta = (hpwlBefore_ == 0.0)
+    ? 0.0
+    : ((double)(hpwlAfter_ - hpwlBefore_) / (double)hpwlBefore_) * 100.;
+  logger_->report("Delta HPWL            {:10.1f} %", hpwl_delta);
+  logger_->report("");
+
+  // Cleanup.
+  if (nw_) delete nw_;
+  if (arch_) delete arch_;
+  if (rt_) delete rt_;
```

Comment:
I don't think these can ever be null here.  In any case it is better to use a std::unique_ptr to avoid manual memory management  (it's a nice upgrade for those of us who have worried about such leaks for years).